File: src/cleansia_android/core/src/main/java/cz/cleansia/core/ui/components/OrderTrackerBar.kt
package cz.cleansia.core.ui.components

import androidx.compose.animation.core.LinearEasing
import androidx.compose.animation.core.RepeatMode
import androidx.compose.animation.core.animateFloat
import androidx.compose.animation.core.infiniteRepeatable
import androidx.compose.animation.core.rememberInfiniteTransition
import androidx.compose.animation.core.tween
import androidx.compose.foundation.background
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxHeight
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Brush
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.layout.layout
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp

/** Cancelled reads red rather than brand — it is the one phase that is not progress. */
private val DangerCancel = Color(0xFFDC2626)

/**
 * Segmented progress bar, one thin bar per phase: past solid, current animating, future muted.
 *
 * **No shimmer, no looping, no halo** — the bar is intentionally quiet so the reader's eye stays on the
 * job, not the chrome.
 *
 * It takes a step INDEX rather than an order status because the two apps do not share a status type:
 * the partner reads the generated `OrderStatus` off its own NSwag client, the customer a hand-written
 * presentation enum. Each maps its own statuses to an index and hands over the already-translated
 * counter label, which also keeps this file free of an `R` it cannot reach from `:core`.
 *
 * @param currentStep 0-based index of the phase in flight. At or past [totalSteps] every segment is drawn as past.
 * @param stepCounterLabel already-formatted and translated, e.g. "Step 3 / 5". Null hides the counter.
 * @param cancelled draws the single full-width red bar instead — a dead order has no progress to show.
 */
@Composable
fun OrderTrackerBar(
    currentStep: Int,
    stepCounterLabel: String?,
    modifier: Modifier = Modifier,
    totalSteps: Int = 5,
    cancelled: Boolean = false,
    cancelledLabel: String? = null,
) {
    if (cancelled) {
        CancelledFullBar(label = cancelledLabel, modifier = modifier)
        return
    }

    // All segments live on the brand palette. Completed and past phases used to read green to signal
    // "done", but a single-color tracker reads cleaner — the past/current distinction comes from
    // saturation, not hue.
    val activeColor = MaterialTheme.colorScheme.primary
    val muted = MaterialTheme.colorScheme.outlineVariant
    val allPast = currentStep >= totalSteps

    // No left-side step name — the headline above the bar already carries it, and repeating it here was
    // visual double-billing. Only the right-aligned counter remains, tucked tight above the bar so the
    // whole tracker reads as a single sub-element under the headline rather than a separate block.
    Column(modifier = modifier.fillMaxWidth()) {
        if (stepCounterLabel != null) {
            Row(
                modifier = Modifier.fillMaxWidth(),
                horizontalArrangement = Arrangement.End,
            ) {
                Text(
                    text = stepCounterLabel,
                    style = MaterialTheme.typography.labelSmall,
                    color = MaterialTheme.colorScheme.onSurfaceVariant,
                )
            }
            Spacer(Modifier.height(4.dp))
        }
        Row(
            modifier = Modifier.fillMaxWidth(),
            horizontalArrangement = Arrangement.spacedBy(6.dp),
        ) {
            repeat(totalSteps) { index ->
                ProgressSegment(
                    state = when {
                        allPast || index < currentStep -> SegmentState.Past
                        index == currentStep -> SegmentState.Current
                        else -> SegmentState.Future
                    },
                    activeColor = activeColor,
                    mutedColor = muted,
                    modifier = Modifier.weight(1f),
                )
            }
        }
    }
}

private enum class SegmentState { Past, Current, Future }

/**
 * One thin pill bar: past solid, future muted, current an indeterminate band sliding across a pale
 * tinted track.
 *
 * The background tint is what stops the current bar reading as inactive between sweeps.
 */
@Composable
private fun ProgressSegment(
    state: SegmentState,
    activeColor: Color,
    mutedColor: Color,
    modifier: Modifier = Modifier,
) {
    val shape = RoundedCornerShape(2.dp)

    when (state) {
        SegmentState.Past -> Box(
            modifier = modifier
                .height(4.dp)
                .clip(shape)
                .background(activeColor),
        )
        SegmentState.Future -> Box(
            modifier = modifier
                .height(4.dp)
                .clip(shape)
                .background(mutedColor),
        )
        SegmentState.Current -> {
            val transition = rememberInfiniteTransition(label = "currentSegmentSweep")
            // A single sweep with a deliberate rest beat: the phase travels past 1f so the band sits
            // off-screen for a pause before wrapping, rather than restarting the instant it exits.
            val sweepWidthFraction = 0.5f
            // Phase range: the visible portion is [-0.5, 1.0] (1.5 units) and the rest beat is
            // [1.0, 2.5] (1.5 units). Equal halves means the sweep is visible for half the cycle and the
            // segment sits in its armed-track tint for the other half — at 1300ms that is a ~650ms sweep
            // and a ~650ms breather between passes.
            val travelEnd = 2.5f
            val phase by transition.animateFloat(
                initialValue = -sweepWidthFraction,
                targetValue = travelEnd,
                animationSpec = infiniteRepeatable(
                    animation = tween(durationMillis = 1300, easing = LinearEasing),
                    repeatMode = RepeatMode.Restart,
                ),
                label = "currentSegmentPhase",
            )

            // The sweep is a horizontal gradient fading transparent → very soft activeColor across its
            // width. Peak alpha is kept low (~0.30) so the band reads as a faint glow gliding over the
            // armed track instead of a saturated wipe — quiet enough to live on screen for hours.
            val sweepBrush = Brush.horizontalGradient(
                colors = listOf(
                    activeColor.copy(alpha = 0f),
                    activeColor.copy(alpha = 0.18f),
                    activeColor.copy(alpha = 0.30f),
                ),
            )

            Box(
                modifier = modifier
                    .height(4.dp)
                    .clip(shape)
                    // Armed-track tint behind the sweep so the bar never looks dead between passes.
                    .background(activeColor.copy(alpha = 0.22f)),
            ) {
                Box(
                    modifier = Modifier
                        .fillMaxWidth(fraction = sweepWidthFraction)
                        .fillMaxHeight()
                        .offsetXFraction(phase)
                        .background(sweepBrush),
                )
            }
        }
    }
}

/**
 * Offsets a child horizontally by [fraction] of its parent's width. Negative fractions push left of the
 * parent, > 1 push right past it. Compose has no built-in fractional offset modifier, so this layout
 * pass does it directly.
 */
private fun Modifier.offsetXFraction(fraction: Float): Modifier = this.layout { measurable, constraints ->
    val placeable = measurable.measure(constraints)
    layout(placeable.width, placeable.height) {
        val parentWidth = constraints.maxWidth
        placeable.placeRelative(
            x = (parentWidth * fraction).toInt(),
            y = 0,
        )
    }
}

@Composable
private fun CancelledFullBar(label: String?, modifier: Modifier = Modifier) {
    Column(modifier = modifier.fillMaxWidth()) {
        if (label != null) {
            Row(
                modifier = Modifier.fillMaxWidth(),
                verticalAlignment = Alignment.CenterVertically,
                horizontalArrangement = Arrangement.SpaceBetween,
            ) {
                Text(
                    text = label,
                    style = MaterialTheme.typography.labelLarge.copy(fontWeight = FontWeight.SemiBold),
                    color = DangerCancel,
                )
            }
            Spacer(Modifier.height(8.dp))
        }
        Box(
            modifier = Modifier
                .fillMaxWidth()
                .height(4.dp)
                .clip(RoundedCornerShape(2.dp))
                .background(DangerCancel),
        )
    }
}

File: src/cleansia_ios/CleansiaCore/Sources/CleansiaCore/Components/OrderTrackerBar.swift
import SwiftUI

/// One phase of the five-phase tracker.
public enum OrderTrackerSegment: Equatable {
    case past, current, future
}

/// What the bar draws: five segments with a step number, or the single flat bar a cancelled order gets.
///
/// One sealed answer rather than a segment list plus a cancelled flag, so a renderer cannot forget the
/// cancelled branch.
public enum OrderTrackerState: Equatable {
    case cancelled
    case steps(segments: [OrderTrackerSegment], stepNumber: Int)
}

/// The phase→segments rule, shared by both apps.
///
/// It takes a step INDEX rather than an order status because the two apps do not share a status type:
/// each generates its own from its own API contract. Each maps its statuses to an index and hands over
/// the already-translated labels.
public enum OrderTrackerRule {
    public static let stepCount = 5

    public static func state(step: Int, cancelled: Bool, completed: Bool) -> OrderTrackerState {
        guard !cancelled else { return .cancelled }
        let index = max(0, min(step, stepCount - 1))
        let segments = (0 ..< stepCount).map { position -> OrderTrackerSegment in
            if completed || position < index { return .past }
            return position == index ? .current : .future
        }
        return .steps(segments: segments, stepNumber: index + 1)
    }
}

enum TrackerMetrics {
    static let barHeight: CGFloat = 4
    static let corner: CGFloat = 2
    static let gap: CGFloat = 6
    static let counterGap: CGFloat = 4
    static let bandWidthFraction: CGFloat = 0.5
    /// The band enters from off-screen left and finishes well past the right edge, so half of each
    /// cycle is a rest beat on the armed track.
    static let travelStart: CGFloat = -0.5
    static let travelEnd: CGFloat = 2.5
    static let cycleSeconds: Double = 1.3
}

/// Segmented progress bar, one thin bar per phase: past solid, current sweeping, future muted.
///
/// **No shimmer, no looping, no halo** — the bar is intentionally quiet so the reader's eye stays on the
/// job, not the chrome. Completed fills every segment rather than sitting on the last; cancelled drops
/// the segmentation entirely, because a dead order has no progress to show.
///
/// Twin: `OrderTrackerBar` (Android, `:core`). Keep the two in step.
public struct OrderTrackerBar: View {
    private let state: OrderTrackerState
    private let stepCounterLabel: (Int, Int) -> String
    private let cancelledLabel: String

    public init(
        state: OrderTrackerState,
        cancelledLabel: String,
        stepCounterLabel: @escaping (Int, Int) -> String
    ) {
        self.state = state
        self.stepCounterLabel = stepCounterLabel
        self.cancelledLabel = cancelledLabel
    }

    public var body: some View {
        switch state {
        case .cancelled:
            CancelledTrack(label: cancelledLabel)
        case let .steps(segments, stepNumber):
            VStack(alignment: .trailing, spacing: TrackerMetrics.counterGap) {
                Text(stepCounterLabel(stepNumber, OrderTrackerRule.stepCount))
                    .font(CleansiaTypography.labelSmall)
                    .foregroundColor(CleansiaColors.onSurfaceVariant)
                HStack(spacing: TrackerMetrics.gap) {
                    ForEach(Array(segments.enumerated()), id: \.offset) { _, segment in
                        TrackerSegment(state: segment)
                    }
                }
            }
            .frame(maxWidth: .infinity, alignment: .trailing)
        }
    }
}

private struct TrackerSegment: View {
    let state: OrderTrackerSegment

    var body: some View {
        switch state {
        case .past:
            flatBar(CleansiaColors.primary)
        case .future:
            flatBar(CleansiaColors.outlineVariant)
        case .current:
            SweepingTrackerSegment()
        }
    }

    private func flatBar(_ color: Color) -> some View {
        RoundedRectangle(cornerRadius: TrackerMetrics.corner)
            .fill(color)
            .frame(height: TrackerMetrics.barHeight)
    }
}

/// The current phase: an armed track under a soft band gliding left to right on a linear loop. Quiet by
/// design — it sits on screen for a multi-hour job.
private struct SweepingTrackerSegment: View {
    @Environment(\.accessibilityReduceMotion) private var reduceMotion
    @State private var sweptToEnd = false

    var body: some View {
        GeometryReader { geometry in
            let width = geometry.size.width
            Rectangle()
                .fill(
                    LinearGradient(
                        colors: [
                            CleansiaColors.primary.opacity(0),
                            CleansiaColors.primary.opacity(0.18),
                            CleansiaColors.primary.opacity(0.30)
                        ],
                        startPoint: .leading,
                        endPoint: .trailing
                    )
                )
                .frame(width: width * TrackerMetrics.bandWidthFraction)
                .offset(x: width * (sweptToEnd ? TrackerMetrics.travelEnd : TrackerMetrics.travelStart))
        }
        .frame(height: TrackerMetrics.barHeight)
        .background(CleansiaColors.primary.opacity(0.22))
        .clipShape(RoundedRectangle(cornerRadius: TrackerMetrics.corner))
        .onAppear {
            guard !reduceMotion else { return }
            withAnimation(
                .linear(duration: TrackerMetrics.cycleSeconds).repeatForever(autoreverses: false)
            ) {
                sweptToEnd = true
            }
        }
    }
}

private struct CancelledTrack: View {
    let label: String

    var body: some View {
        VStack(alignment: .leading, spacing: Spacing.xs) {
            Text(label)
                .font(CleansiaTypography.labelLarge)
                .foregroundColor(CleansiaColors.error)
            RoundedRectangle(cornerRadius: TrackerMetrics.corner)
                .fill(CleansiaColors.error)
                .frame(height: TrackerMetrics.barHeight)
        }
        .frame(maxWidth: .infinity, alignment: .leading)
    }
}
